### ISSUE TITLE
Is it possible to get the parameters used for a specific build?

### ISSUE THREAD

**Issue by pgrandin** [salimfadhley/jenkinsapi#438]
When looking at a build from the UI, you can see the parameters that were used for the build.
I cannot find how to fetch that information via jenkinsapi.

I can fetch the parameters of the job itself, but not the ones that were used for the build.

Is this possible to do?

Thanks!

**Comment by lechat** [comment]
Sure!

```
j = Jenkins('http://localhost:8080')
job = j.get_job("foo")
build = job.get_last_build()
parameters = build.get_actions()['parameters'] 
```

parameter variable will have a list of dicts:
```
[{
    'name': 'parameter_name',
    'value': 'parameter_value'
}]
```

Hope this helps.

**Comment by joyceqiao** [comment]
hi,

I have tried but can not get it, here is the error:
Traceback (most recent call last):
  File "getpara.py", line 30, in <module>
    print last_bld.get_actions()['parameters']
KeyError: 'parameters'

my code is here:
      1 #!/bin/env python
      2 from optparse import OptionParser
      3 import sys
      4 import jenkinsapi
      5 from jenkinsapi.jenkins import Jenkins
      6 
      7 parser = OptionParser()
      8 
      9 parser.add_option("-j", "--jobname", dest="job", action="store",
     10                     type="string", help="The job name.")
     11 parser.add_option("-p", "--parameter", dest="para", action="store",
     12                     type="string", help="The parameter of the job")
     13 
     14 (options, args) = parser.parse_args()
     15 
     16 if not options.job or not options.para:
     17   print("Usage:%s -j job -p parameter."  % (__file__))
     18   sys.exit(1)
     19 
     20 JenkinsServer="http://1.1.1.1"
     21 username = "user"
     22 passwd = "***"
     23 jen = Jenkins(JenkinsServer, username, passwd)
     24 
     25 job= jen.get_job(options.job)
     26 last_bld = job.get_last_build()
     27 if last_bld :
     28    print last_bld.get_actions()['parameters']

I'm using  jenkinsapi-0.2.24


Thanks

**Comment by pgrandin** [comment]
Same here, using jenkinsapi-0.2.29. My code:

``from jenkinsapi.jenkins import Jenkins
[...]
j = Jenkins(url, username = username, password = password)
job = j.get_job(job_name)
build = job.get_last_build()
print build.get_actions()
``

I get :

> {'name': 'value', 'causes': [{'userName': 'pierre.grandin', 'userId': 'pierre.grandin', 'shortDescription': 'Started by user pierre.grandin'}]}

And that's it, no parameters. My jenkins server is running 1.653, and I can access the parameters from the UI.

**Comment by lechat** [comment]
Just tested it and I got parameters as expected...

What do you see when you go to the build's json API page?

http://localhost:8080/job/<job_name>/<build_number>/api/json?pretty=true

**Comment by pgrandin** [comment]
In the build's json API page i do see my parameter :+1: 

> {
  "actions" : [
    [
      {
        "name" : "AVAILABILITY_ZONE",
        "value" : "us-east-1b"
      },
      {
        "name" : "vpc",
        "value" : ""
      },
      {
        "name" : "host_to_deploy",
        "value" : "dev-jenkins-slave03"
      },
      {
        "name" : "action",
        "value" : "start"
      },
      {
        "name" : "pupenv",
        "value" : "dev"
      },
    ],
    {
      "causes" : [
        {
          "shortDescription" : "Started by user pierre.grandin",
          "userId" : "pierre.grandin",
          "userName" : "pierre.grandin"
        }
      ]
    },

**Comment by lechat** [comment]
I've tried on Jenkins 2.0 and 1.651.1 (LTS) and in both "actions" has "parameters" element:

```
  "actions" : [
    {
      "parameters" : [
        {
          "name" : "PAR1",
          "value" : "1"
        }
      ]
    },
    {
      "causes" : [
        {
          "shortDescription" : "Started by user anonymous",
          "userId" : null,
          "userName" : "anonymous"
        }
      ]
    }
  ],
```

**Comment by pgrandin** [comment]
Mmm, that's weird. 

The api output I shared gets mapped as 
> {'name': 'value', 'causes': [{'userName': 'pierre.grandin', 'userId': 'pierre.grandin', 'shortDescription': 'Started by user pierre.grandin'}]}

I wonder why my server is missing the parameters object. Maybe we can add a workaround in the api code?

**Comment by lechat** [comment]
I don't think that adding a "workaround" to your specific issue into library is a good idea. You shall make workaround in your code.

**Comment by pgrandin** [comment]
Unfortunately I do not see where to do it in my code, that uses the API.
I do not mind running a version of the library patched just for this specific issue, and i definitely understand that you don't want to do that in the official repo. 
I will dig in the API code to see if I can do it. Any pointers, maybe?

**Comment by lechat** [comment]
`build.get_actions()` gives you access to `actions` dict returned by Jenkins. You can do anything with that dict, isn't it? I am not sure what would you expect library would do in your case.

**Comment by pgrandin** [comment]
Well the dict gets mapped as {'name': 'value'}... Which isn't really useful.. I am not saying that it's an issue in the library, i'm just trying to figure out a way to get this working :)

**Comment by joyceqiao** [comment]
hi,

I cann't get parameters either. My result is the same as @pgrandin 's
{'name': 'value', 'causes': [{'userName': '\xe5\xbc\xa0\xe4\xb8\xb9', 'userId': 'test', 'shortDescription': 'Started by user \xe5\xbc\xa0\xe4\xb8\xb9'}]}

actually, this worked well before I upgrade my jenkins .

**Comment by lechat** [comment]
...and 2.x has parameters again. So, not really `jenkinsapi` problem if something got broken and later fixed in Jenkins.

**Comment by pgrandin** [comment]
Confirmed that we fixed it by downgrading to 1.651.2

Thanks @lechat !

**Comment by akostadinov** [comment]
FYI this is how I get params within a pipeline:
https://gist.github.com/akostadinov/fc8940fda63a7489492b637063233524#file-pipeline_examples-groovy-L17-L25
